### DISCUSSION TITLE
Custom Client Directives

### DISCUSSION THREAD

**Discussion by matthewp** [withastro/roadmap#272] [Proposal]
- Start Date: 2022-08-31
- Reference Issues: https://github.com/withastro/rfcs/discussions/272
- Implementation PR: 

# Summary

Provide an API for integrations to implement custom `client:` directives to provide greater control for when client-side JS is loaded and executed.

# Example

## Usage

To use a custom directive in your app, add an integration like any other.

```js
import { defineConfig } from 'astro/config';
import onClickDirective from '@matthewp/astro-click-directive';

export default defineConfig({
  integrations: [onClickDirective()]
});
```

And then use it in your Astro components:

```astro
---
import Counter from '../components/Counter.jsx';
---

<Counter client:click />
```

## Implementation

An implementer of a custom client directive would do so through the hooks API:

```js
export default function() {
  return {
    hooks: {
      'astro:config:setup': ({ addClientDirective }) => {
        addClientDirective({
          name: '@matthewp/astro-click-directive',
          key: 'click',
          entrypoint: '@matthewp/astro-click-directive/client.js'
        });
      },
    }
  }
}
```

The entrypoint being:

```js
export default function(load, opts, element) {
  element.addEventListener('click', async () => {
    const hydrate = await load();
    await hydrate();
  }, { once: true });
}
```

# Motivation

The last client directive added to core was the `client:only` directive in [August 2021](https://github.com/withastro/astro/issues/751). Since that time the core team has been hesitant to add new client directives despite the community asking about them.

Allowing custom client directives would both:

- Allow the community to experiment with different approaches to lazy-loading client JavaScript.
- Provide evidence, through telemetry data, on which directives are most used. This data could be used to determine if a directive should be brought into core.

Some examples of custom directives that people have wanted in the past:

- Loading JavaScript on client interactive, such as mouseover or click.
- Loading JavaScript when an element is visible, as opposed to within the viewport as `client:visible` currently does.
- The [Idle Until Urgent](https://philipwalton.com/articles/idle-until-urgent/) pattern which loads on either idle or interaction, whichever comes first.

# Detailed design

## Client Directive API

A client directive is a file that exports a function that handles loading and hydrating components. The type is:

```ts
type Hydrate = () => Promise<void>;
type Load = () => Promise<Hydrate>;

type DirectiveOptions = {
  // The component displayName
  name: string;
  // The attribute value provided,
  // for ex `client:interactive="click"
  value: string;
}

export type ClientDirective = (load: Load, opts: DirectiveOptions, element: HTMLElement);
```

A directive might import and use this type like so:

```ts
import type { ClientDirective } from 'astro';

const directive: ClientDirective = (load, opts, element) => {
  element.addEventListener('click', async () => {
    const hydrate = await load();
    await hydrate();
  }, { once: true });
};

export default directive;
```

## Integration API

The API for registering a custom directive is modelled on the `addRenderer` and `setAdapter` APIs, where you provide a name for the directive and a path to the implementation.

```js
export default function() {
  return {
    hooks: {
      'astro:config:setup': ({ addClientDirective }) => {
        addClientDirective({
          name: '@matthewp/astro-click-directive',
          key: 'click',
          entrypoint: '@matthewp/astro-click-directive/client.js'
        });
      },
    }
  }
}
```

Similar to how renderers and adapters work, the directive will be stored on the configuration's context object as a key/value pair of directive keys (such as `click` in this example) and the directive options (the object shown in this example).

This means that if there are conflicting directives; directives that are using the same key, only one will "win", which will be whichever comes last.

Our own builtin core directives *will also* use this pattern, and this means that it will be possible for a custom directive to __override__ an existing built-in directive; meaning you can redefine `client:visible` to do something different.

## Implementation

Currently all client directives are built into Astro and injected onto the page as a string as JavaScript. We want to replicate this for production builds in order to minimize the number of external requests and prevent waterfalls.

During development a directive will be injected like so:

```html
<script type="module" src="/@id/@matthewp/astro-click-directive/client.js?astro-client-directive"></script>
```

A Vite plugin will define a virtual module that looks like this:

```js
import directive from '@matthewp/astro-click-directive/client.js';

(self.Astro = self.Astro || {}).click = directive;
window.dispatchEvent(new Event('astro:click'));
```

During build type this will get built and bundled and be injected as an inline script in production that looks like:

```html
<script>
"use strict";
{
  const directive = (load, opts, element) => {
    element.addEventListener('click', async () => {
      const hydrate = await load();
      await hydrate();
    }, { once: true });
  };

  (self.Astro = self.Astro || {}).click = directive;
  window.dispatchEvent(new Event('astro:click'));
}
</script>
```

# Drawbacks

- Directives are global and not imported. Someone coming into the project might not know what `client:foo` means since they can't look it up in our documentation.
- It's possible that someone could create a directive that is a lot of code and causes the page to load slower.
  - We could set some file-size limit for directives and error if any exceed it.
- Implementation will require a small refactor of how directives work, but this refactor will improve the code.
- Making this extensible might cause core to deprioritize builtin directives.
- Having a custom directive override an existing built-in one can cause confusion.

# Adoption strategy

- This change will be non-breaking.
- This was designed in such a way that custom directives always override built-in directives. This means that if core adds a new directive it will *not* interfere with a custom directive of the same name; the user's code will continue to work the same.

**Comment by elbotho** [comment]
I'd love this. I think there are some directives that would make a lot of sense.

I actually started work on a custom directive on focus or other dom events but was intimidated to work inside the core.
So a clearly defined API like this would make this a lot easier.

**Comment by Princesseuh** [comment]
Pff! It's dusty in here, been a while. 

As we've recently revamped our roadmap process, we'd like to circle back to some earlier proposals that did eventually become PRs, but were not fully accepted (as in, they were not merged). This is one of them. The related full PR is here: <https://github.com/withastro/roadmap/pull/278>, but as this is circling back to a Stage 1, we'd like for this to be the place to discuss the proposal.

I'd like to champion pushing this proposal forward, as I'm very jealous of the newer frameworks having more client directives (a thing we invented!) than us.

---

My thoughts: This is very good! Also very in line with other additions to Astro we support at the moment (integrations, adapters, remark plugins etc)

The only part I'm a bit meh on is the following:

> Our own builtin core directives *will also* use this pattern, and this means that it will be possible for a custom directive to __override__ an existing built-in directive; meaning you can redefine `client:visible` to do something different.

Overriding Astro defaults has always been a very painful point, how many times have we changed the behaviour around remark / rehype plugins order and overriding of the default ones? As such, I'd say we should **not** allow this. Unlike remark / rehype plugins, having more client directives isn't something negative, apart from the name squatting.

Apart from that, I don't have much to say, very excited about this!

**Reply by matthewp** [reply]
Yeah, the reason I put that is because it would be extra work for us to block people from overriding the directive (this would just be a simple Map), we'd have to have like a denylist or something (but then give our internal usage a secret key to bypass it).

**Comment by matthewp** [comment]
I'm still in favor of this! The reason why I didn't push for it earlier was that the response wasn't overwhelming. I think people felt generally positive but there also wasn't many (any?) people specifically asking for it. I think that's still true.

One thing that this would unlock is https://github.com/withastro/roadmap/discussions/377; this would give us a single place where we know all things that might be inlined, so we could hash them and provide that those hashes for those who need to avoid `unsafe-inline`. That's a completely separate discussion, of course, just pointing out how this refactor would help with that.

**Comment by bluwy** [comment]
This looks great to me too. Even without many asking for this feature, I think implementing this would open up more user optimization opportunities. The trickiest part might be getting autocompletion to work, but I bet it's doable at a later stage.

**Reply by matthewp** [reply]
@Princesseuh Can vscode read any config files in order to know what directives exist? If so that could be a solution to this part.

**Reply by Princesseuh** [reply]
> @Princesseuh Can vscode read any config files in order to know what directives exist? If so that could be a solution to this part.

It could, or we could ask user to make / generate TypeScript interfaces and get the list from that + type checking

**Reply by matthewp** [reply]
Ah, yeah if we can do d.ts files then that's the way to go.

**Comment by amxmln** [comment]
I’d be interested in this as well! 😊 It would allow something like I outlined in https://github.com/withastro/roadmap/discussions/497 and even more complex cases on top of that.

**Comment by delucis** [comment]
Would this model also support custom `client:only` directives as requested here for example? https://github.com/withastro/roadmap/discussions/272

I think that's the one directive that's a bit special because it turns off prerendering the component but still having JS load timing control would be nice and not available currently.

**Reply by matthewp** [reply]
Not sure I totally understand? `client:only` does require some special compiler integration, so it's not something that would be possible in userland, if that's the question. All of the others would, though.

This proposal just gives you a hook to decide when the code should be loaded in the client. Perhaps we *could* extend it so that you are given some hooks for the server-side part too. But I'd want to see a use-case for that.

**Reply by delucis** [reply]
I meant that currently you can't have `client:only` components hydrate on idle/visible, you just get the basic behaviour which I guess is hydration on page load? (Exactly how `visible` works for a component that hasn't rendered yet TBD.)

**Reply by matthewp** [reply]
Ok, I understand now. You're really pointing how that `client:` has dual meaning. In most cases it means "when to load" but `client:only` is actually about telling the server *not* to load. What if it was instead named `server:none`, then you'd be able to do: `<Component server:none client:idle />`. 

This aligns with @FredKSchott's idea of a `client:none` directive, to tell a framework component not to hydrate (which is of course the current default with no directives).

**Reply by preetamslot** [reply]
This would be quite helpful. and the `server:none client:idle` sounds like a great solution.

**Reply by carvinlo** [reply]
@matthewp Sorry to bother you, has the `server:none` been implemented?

**Reply by delucis** [reply]
@carvinlo No, that was just an idea.

**Reply by carvinlo** [reply]
It would be more perfect if it could disable server-side rendering and use visible, idle and other direcitves together. Am I missing something?

**Reply by delucis** [reply]
Yes, that’s what the idea was trying to address. It could indeed be helpful in some circumstances! It just hasn’t actually been implemented yet.

**Comment by matthewp** [comment]
This has been accepted and moved to stage 2: https://github.com/withastro/roadmap/issues/529
